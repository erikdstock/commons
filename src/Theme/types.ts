import { SystemStyleObject, Theme, ThemeValue } from "@styled-system/css"
import * as CSS from "csstype"

type ObjectOrArray<T> = T[] | { [K: string]: T | ObjectOrArray<T> }

interface BaseThemeKnownColors {
  text: CSS.ColorProperty
  background: CSS.ColorProperty
  primary: CSS.ColorProperty
  secondary: CSS.ColorProperty
  muted: CSS.ColorProperty
}

interface Variants {
  variants?: {
    [k: string]: SystemStyleObject
  }
}

type SystemStyleWithVariants = SystemStyleObject & Variants

interface RosesThemeColors extends BaseThemeKnownColors {
  [k: string]: CSS.ColorProperty | ObjectOrArray<CSS.ColorProperty>
}
interface RosesThemeFonts {
  // heading: ThemeValue<CSS.FontFamilyProperty>
  // text: ThemeValue<CSS.FontFamilyProperty>
  [k: string]: ThemeValue<CSS.FontFamilyProperty>
}
/** The roses theme follows the styled-system/system-ui theme specification */
export type RosesThemeObject = Theme & {
  fonts?: RosesThemeFonts
  colors: RosesThemeColors
  /** Theme-ui html styles (incoming from baseTheme) */
  styles?: { [elementName: string]: SystemStyleWithVariants }
  /** Roses html styles */
  htmlStyles?: { [elementName: string]: SystemStyleObject }
  /** Roses component styles */
  componentStyles?: { [componentName: string]: SystemStyleWithVariants }
  /** Custom component variant styles */
}
