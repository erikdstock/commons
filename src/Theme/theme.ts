import { Theme } from "styled-system"

const colors = {
  red: "#db1d33",
  black: "#000",
  white: "#fff",
  gray: {
    "1": "hsl(0,0%,32%)",
    "2": "hsl(0,0%,47%)",
    "3": "hsl(0,0%,58%)",
    "4": "hsl(0,0%,68%)",
    "5": "hsl(0,0%,77%)",
    "6": "hsl(0,0%,85%)",
    "7": "hsl(0,0%,93%)",
  },
}

// Default Breakpoints
const breakpoints = ["40em", "52em", "64em"]

// default fontSizes
const fontSizes = [12, 14, 16, 20, 24, 32, 48, 64, 72]

// default space for margin and padding
const space = [0, 4, 8, 16, 32, 64, 128, 256, 512]

// type NestedObjectOrArray<T> = T[] | { [K: string]: T | NestedObjectOrArray<T> }

interface CommonsTheme extends Theme {
  // colors: NestedObjectOrArray<CSS.ColorProperty>
  colors: any
}

export const defaultTheme: CommonsTheme = {
  breakpoints,
  colors,
  fontSizes,
  space,

  fonts: {
    serif: "'IBM Plex Serif', serif",
    sans: "Chivo, Helvetica, Sans",
    mono: "'Fira Mono', monospace",
  },

  radii: [0, 2, 3, 5, 10],
  buttons: {
    primary: {
      color: colors.white,
      backgroundColor: colors.red,
    },
    outline: {
      color: colors.red,
      backgroundColor: colors.white,
      boxShadow: "inset 0 0 0 2px",
    },
    inactive: {
      backgroundColor: colors.gray[5],
      color: colors.white,
      disabled: true,
    },
  },
}
