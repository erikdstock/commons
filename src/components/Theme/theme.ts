import { NONAME } from "dns";

/**
 * An interface matching what what rebass expects.
 * See https://rebassjs.org/theming
 */
export interface RebassTheme {
  /* Array of strings representing viewport widths to use for min-width media queries. */
  breakpoints?: string[]
  
  /* Array of numbers to use as a typographic scale */
  fontSizes?:	[number|string]
  
  /* Color names that can be used in color, bg, and borderColor props (nested objects can be accessed via dot notation) */
  colors?: object
  
  /* Array of numbers for use as margin and pixel values */
  space?: object

  /* Values for the fontFamily prop */
  fonts?: object
  
  /* Array or Object values for fontWeight prop */
  fontWeights?: object

  /* Array or Object	Values for lineHeight prop */
  lineHeights?: object

  /* 	Array or Object	Values for letterSpacing prop */
  letterSpacings?: object

  /* Array or Object	Values for boxShadow prop */
  shadows?:	object

  /* Array or Object	Values for border props */
  borders?: object

  /* Array or Object	Values for borderRadius props */
  radii?: object

  /* Array or Object	Values for opacity props */
  opacity?: object

  /* Styles for button varians */
  buttons: object
}

const colors = {
  red: "#db1d33",
  black: "#000",
  white: "#fff",
  gray: {
    "1": "hsl(0,0%,32%)",
    "2": "hsl(0,0%,47%)",
    "3": "hsl(0,0%,58%)",
    "4": "hsl(0,0%,68%)",
    "5": "hsl(0,0%,77%)",
    "6": "hsl(0,0%,85%)",
    "7": "hsl(0,0%,93%)",
  }
}

export const defaultTheme: RebassTheme = {
  fonts: {
    serif: "'IBM Plex Serif', serif",
    sans: "Chivo, Helvetica, Sans",
    mono: "'Fira Mono', monospace"
  },

  colors,
  buttons: {
    primary: {
      color: colors.white,
      backgroundColor: colors.red,
    },
    outline: {
      color: colors.red,
      backgroundColor: colors.white,
      boxShadow: 'inset 0 0 0 2px'
    },
    inactive: {
      backgroundColor: colors.gray[5],
      color: colors.white,
      disabled: true
    },
  }
}