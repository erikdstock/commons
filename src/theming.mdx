---
Theming
---
import { Box } from './components/Layout'
import { H2 } from './components/Typography'

# Theming

<Box py={3} px={6} bg="red" aligngItems="center" color="white"><H2 fontFamily="mono" color="white">TODO</H2></Box>