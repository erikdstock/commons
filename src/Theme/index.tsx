import React from "react";
import { ThemeProvider } from "styled-components";
import { defaultTheme, RebassTheme } from "./theme";

export { defaultTheme, RebassTheme };

interface Props {
  theme?: RebassTheme;
}


/**
 * A theme provider
 */
export const Theme: React.FunctionComponent<Props> = ({
  children,
  theme = defaultTheme
}) => (
  <ThemeProvider theme={theme}>
    <>
      {
        /* ThemeProvider Children must be a node, so Fragment </> is required */
        children
      }
    </>
  </ThemeProvider>
);
