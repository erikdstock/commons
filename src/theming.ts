import styled from "@emotion/styled"
import styledCss, { CSSObject } from "@styled-system/css"
import get from "lodash/get"
import React from "react"
import { RosesThemeObject } from "./Theme"
import { RosesSC, RosesStyledComponent, RosesStyleProps } from "./types"
import isPropValid from "@emotion/is-prop-valid"

/** The base style for Box and everything on up. */
export const baseStyle: CSSObject = {
  // Type assertion is required here so that type
  // will not be widened by ts compiler to { boxSizing: string }
  boxSizing: "border-box" as "border-box",
}

type InterpolationFn = (
  p: RosesStyleProps & { theme: RosesThemeObject }
) => CSSObject | undefined

const componentStylesRoot: keyof RosesThemeObject = "componentStyles"

const validHtmlPropChecker = (prop: string) => {
  return isPropValid(prop)
}

/**
 * Add `sx` + `variant` props, get theme values
 * @param name The key under componentStyles where this component's styles will be located
 * @param as The type of element to return
 * // TODO: Test
 */
export const simpleThemed = (
  name: string,
  as: keyof JSX.IntrinsicElements = "div"
): RosesSC => {
  return styled<any, RosesStyleProps>(as, {
    shouldForwardProp: validHtmlPropChecker,
  })(baseStyle, relevantStyles(`${componentStylesRoot}.${name}`), sxHandler)
}

/** Because @styled-system/css returns a theme-ready function,
 *   but we are in that function now, just call it right away.
 *   This way we don't interpolate an additional function for
 *   emotion to execute.
 */
export const sxHandler: InterpolationFn = ({ sx, ...restProps }) => {
  return sx && styledCss(sx)(restProps)
}

/** Find styles, extract and apply them + variants using @styled-system/css */
// TODO: Refactor to allow passing styles + variants directly rather than putting them in the theme?
const relevantStyles = (themePath: string): InterpolationFn => props => {
  const { theme, variant } = props
  const { variants, ...styles } = get(theme, themePath, {})
  if (styles) {
    let selectedVariant
    if (variant && variants) {
      selectedVariant = variants[variant]
    }
    return styledCss({ ...styles, ...selectedVariant })(props)
  } else return {}
}

interface ThemedComponentOptions {
  name: string
  shouldForwardProp?: typeof isPropValid
  defaultVariant?: string
  component?: React.ComponentType<any> | keyof JSX.IntrinsicElements
}

/**
 * Wrap a base styled component with:
 *   - a responsive, theme-aware style `sx` prop
 *   - a `variant` prop for accessing styles set on the theme.variants key
 * @example:
 *
 *  const Box = themed({name: 'Box', component: RawBox, defaultVariant: "hot"})
 *  const Widget = themed('Widget')
 *
 */
export const themed = (
  options: ThemedComponentOptions | string
): RosesStyledComponent => {
  if (typeof options === "string") {
    const name = options as keyof JSX.IntrinsicAttributes
    return simpleThemed(name)
  } else {
    const {
      component,
      shouldForwardProp = (prop: string) => prop !== "sx" && prop !== "variant",
      name,
    } = options

    return component
      ? styled<React.ComponentType<any>, RosesStyleProps>(
          component as React.ComponentType<any>,
          { shouldForwardProp }
        )(relevantStyles(`${componentStylesRoot}.${name}`), sxHandler)
      : simpleThemed(`${componentStylesRoot}.${name}`)
  }
}
