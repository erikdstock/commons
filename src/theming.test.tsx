/** @jsx jsx */
import { jsx } from "@emotion/core"
import { themed } from "./theming"
import { RosesTheme } from "./Theme"
import renderer from "react-test-renderer"
import { baseTheme } from "./Theme/baseTheme"
import { matchers } from "jest-emotion"

const { colors: baseColors } = baseTheme as any
const theme = {
  ...baseTheme,
  colors: {
    ...baseColors,
    primary: "red",
    secondary: "green",
    muted: "tomato",
  },
}

expect.extend(matchers)

describe("themed()", () => {
  const localTheme = {
    ...theme,
    componentStyles: {
      Widget: {
        color: "primary",
        variants: {
          experimental: {
            color: "secondary",
          },
        },
      },
    },
  }

  it("adds component themes + variants from the theme", () => {
    const DivWidget = themed("Widget")
    const tree = renderer.create(
      <RosesTheme theme={localTheme}>
        <DivWidget />
      </RosesTheme>
    )
    expect(tree.toJSON()).toMatchSnapshot()

    const tree2 = renderer.create(
      <RosesTheme theme={theme}>
        <DivWidget variant="experimental" />
      </RosesTheme>
    )
    expect(tree2.toJSON()).toMatchSnapshot()
  })

  it("has an sx prop for theme-aware styles", () => {
    const DivWidget = themed("Widget")
    const tree = renderer.create(
      <RosesTheme theme={localTheme}>
        <DivWidget sx={{ color: "muted" }} />
      </RosesTheme>
    )
    expect(tree.toJSON()).toMatchSnapshot()
  })
})
