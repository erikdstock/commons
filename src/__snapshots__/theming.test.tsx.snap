// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`simpleThemed() makes a themed div by default 1`] = `
.emotion-0 {
  box-sizing: border-box;
  color: red;
}

<div
  className="emotion-0"
/>
`;

exports[`simpleThemed() makes a themed other thing if you pass that tag in 1`] = `
.emotion-0 {
  box-sizing: border-box;
}

<header
  className="emotion-0"
>
  Welcome to website.
</header>
`;

exports[`themed() adds component themes + variants from the theme 1`] = `
.emotion-0 {
  box-sizing: border-box;
  color: red;
}

<div
  className="emotion-0"
/>
`;

exports[`themed() adds component themes + variants from the theme 2`] = `
.emotion-0 {
  box-sizing: border-box;
}

<div
  className="emotion-0"
/>
`;

exports[`themed() has an sx prop for theme-aware styles 1`] = `
.emotion-0 {
  box-sizing: border-box;
  color: red;
  color: tomato;
}

<div
  className="emotion-0"
/>
`;
